juju/application.py

```diff
@@ -331,7 +331,49 @@ def upgrade_charm(
         :param str switch: Crossgrade charm url
 
         """
-        raise NotImplementedError()
+        # TODO: Support local upgrades
+        if path is not None:
+            raise NotImplementedError("path option is not implemented")
+        if resources is not None:
+            raise NotImplementedError("resources option is not implemented")
+
+        if switch is not None and revision is not None:
+            raise ValueError("switch and revision are mutually exclusive")
+
+        client_facade = client.ClientFacade.from_connection(self.connection)
+        app_facade = client.ApplicationFacade.from_connection(self.connection)
+
+        if switch is not None:
+            charm_url = switch
+        else:
+            charm_url = self.data['charm-url']
+            charm_url = charm_url.rpartition('-')[0]
+            if revision is not None:
+                charm_url = "%s-%d" % (charm_url, revision)
+            else:
+                charmstore = self.model.charmstore
+                entity = await charmstore.entity(charm_url, channel=channel)
+                charm_url = entity['Id']
+
+        await client_facade.AddCharm(
```

Comment:
I just tested this scenario, and the result was that upgrade_charm completed without error.

The Juju CLI, however, raises an error in the same scenario:
```
$ juju deploy cs:~containers/easyrsa
Located charm "cs:~containers/easyrsa-9".
Deploying charm "cs:~containers/easyrsa-9".

$ juju upgrade-charm easyrsa --revision 9
ERROR already running specified charm "cs:~containers/easyrsa-9"
```

I'll add a check here and raise an exception accordingly. Thanks for the feedback!